Added Favicon to Django API

madewithkode opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #4341 by @sarayourfiend

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
Added Favicon to Django API

<img width="1439" alt="Screenshot 2024-05-19 at 15 00 41" src="https://github.com/WordPress/openverse/assets/25867872/260d0ebf-c01a-4f86-93ec-bb74ec196d7d">



## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
- Start the API server -> `just a`

- Navigate to `localhost:50280`

You should be able to observe an Openverse Favicon present on the tab.


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

krysal reviewed (commented):
Thank you for tacking on this, @madewithkode! This looks like a good solution, though after searching in the [Spetacular documentation](https://drf-spectacular.readthedocs.io/en/latest/settings.html#example-swaggerui-settings), there seems to be a simpler approach. It should be enough specifying the favicon URL in the `SWAGGER_UI_FAVICON_HREF` configuration (see the previous link).

I'll get back to you with what URL to use, I think we'll avoid using static files via Django.

madewithkode commented:
Thanks for pointing this out @krysal. Certainly easier and more straightforward lol. Would be standing by for the link!
Also, does Spectacular related config changes require any special file regeneration commands or steps in order for them to take effect or they're picked up in real time?

krysal commented:
@madewithkode Forget what I said about the file location/URL. It should be good there in the `static` folder!

> Also, does Spectacular related config changes require any special file regeneration commands or steps in order for them to take effect or they're picked up in real time?

I don't think running a command is necessary, but @dhruvkb can confirm with us since he knows more about Spectacular.

dhruvkb commented on api/api/static/favicon.svg:
```diff
@@ -0,0 +1,16 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+<rect width="16" height="16" rx="2" fill="#232323"/>
+<g clip-path="url(#clip0_4508_11559)">
```
```suggestion
```

dhruvkb commented on api/api/static/favicon.svg:
```diff
@@ -0,0 +1,16 @@
+<path d="M10.5693 11.9985C11.5568 11.9985 12.3574 11.1977 12.3574 10.2099C12.3574 9.22209 11.5568 8.42133 10.5693 8.42133C9.58179 8.42133 8.78125 9.22209 8.78125 10.2099C8.78125 11.1977 9.58179 11.9985 10.5693 11.9985Z" fill="white"/>
+<path d="M6.19067 10.2069C6.19067 11.1922 6.98474 11.9974 7.9674 11.9974V8.42133C6.9897 8.42133 6.19067 9.22157 6.19067 10.2069Z" fill="white"/>
+<path d="M3.59998 10.2069C3.59998 11.1972 4.39404 11.9974 5.3767 11.9974V8.42133C4.39404 8.42133 3.59998 9.22157 3.59998 10.2069Z" fill="white"/>
+</g>
+<defs>
+<clipPath id="clip0_4508_11559">
+<rect width="8.8" height="8" fill="white" transform="translate(3.59998 4)"/>
+</clipPath>
+</defs>
```
```suggestion
```

dhruvkb reviewed (changes requested):
@krysal is right that `SWAGGER_UI_FAVICON_HREF` should be the approach here. I made the following change to the `api/conf/settings/spectacular.py` file and the favicon showed up immediately (this diff also includes a fix for the broken logo in the API docs).

<img width="922" alt="image" src="https://github.com/WordPress/openverse/assets/16580576/5e382994-d1f3-40c8-8afa-0c7382055739">

With this change^, you should drop the `redoc.html` template and undo the changes to `openapi.py`.

sarayourfriend commented:
This is also missing (and I believe not addressed by the feedback from Krystle and Dhruv) the favicon on the actual API routes themselves. https://api.openverse.org/v1/images/ exists outside the DRF spectacular view, so won't be covered by the swagger changes, I think.

For that, we'll need to override the `rest_framework/api.html` template: https://github.com/encode/django-rest-framework/blob/master/rest_framework/templates/rest_framework/api.html

Documentation about that here: https://www.django-rest-framework.org/topics/browsable-api/#customizing

After extending the base template, [`load static`](https://docs.djangoproject.com/en/5.0/ref/templates/builtins/#std-templatetag-static), open the [`head` block](https://github.com/encode/django-rest-framework/blob/master/rest_framework/templates/rest_framework/base.html#L8), and add `<link rel="shortcut icon" href="{%  static 'favicon.ico' %}">` (and one for SVG for browsers that support it, more on that below) to the `head` block along with calling `super()`.

One last thing: SVG favicons are not supported by all browsers, notably Safari on iOS, even on the latest version: https://caniuse.com/link-icon-svg

I discussed this in the PR that updated the frontend favicon, but we need to supply both: https://github.com/WordPress/openverse/pull/3724

You can generate a multi-size ico from an svg using `svg-to-ico`, but I found I needed to do some manual editing of the ico file in GIMP to remove sizes we didn't have before, and to ensure the size of the file was not unnecessarily huge. This was probably the most complex and definitely the most tedious part of the issue (and hence why I probably wouldn't necessarily recommend this work as a good first issue, by the way, @dhruvkb, in reference to your suggestion on the issue). At the very least, it's a requirement that wasn't spelled out in the issue, and only evident through context of the other recent favicons in our codebase.

madewithkode commented:
Hi @dhruvkb/@krysal  I followed your suggestions of using `spectacular-drf`s `SWAGGER_UI_FAVICON_HREF` and sadly I couldn't quite get it to work. Tried dropping the image and the volumes, repeatedly rebuilt, all with no luck. However the modifications to the logo url worked and the image was displayed. A look at the api web container logs kept showing that as against the specified `favicon.svg` the request was indeed looking for `favicon.ico` which was weird.

Also @sarayourfriend, for individual API pages outside of `drf-spectacular` context, I tried following your suggested approach by overriding `rest_framework/api.html` and `{% block head %} with the links to the favicon as described in the docs and also tried calling `{% block.super %}` but without any luck. I don't know if I am missing something. Regarding generating a multi-size ico from an svg and doing some other modifications with gimp, like you mentioned, I'm not so sure how to go about those, is this something that can be done and the final blobs shared with me somehow?

In the meantime I have updated the PR with the latest changes I tried and also converted it to draft as it is obviously not ready and might require a few more iterations. Looking forward to your insights.

dhruvkb commented:
I have used https://realfavicongenerator.net/ in the past to generate a whole bunch of favicons (and related images). Maybe consider trying that to process the SVG into ICO (and other useful formats as needed for different browsers and platforms).

`SWAGGER_UI_FAVICON_HREF` should've worked (even without rebuilding any of the Docker images). I can see it working just fine locally. Did you drop your custom view and template before setting `SWAGGER_UI_FAVICON_HREF`? Let me pull your PR locally and verify.

madewithkode commented:
Hi @dhruvkb, Yes, I indeed dropped the custom view and templates but no luck. I updated the branch with my latest local state so you should be able to test locally, if it works for you then maybe something is wrong with my local environment.
Also, thanks for recommending  https://realfavicongenerator.net/, I'd have a look.

sarayourfriend commented on api/api/templates/rest_framework/api.html:
```diff
@@ -0,0 +1,5 @@
+{% extends "rest_framework/base.html" %}
+
+{% block head %}
+<link rel="icon" type="image/x-icon" href="{% static 'favicon.ico' %}">
+{% endblock %}
```
[favicon_package_v0.16.zip](https://github.com/WordPress/openverse/files/15409607/favicon_package_v0.16.zip)

Here's the generated favicon package from the website Dhruv recommended. You just need the favicon.ico file (and the original SVG), everything else is unnecessary for the API.

madewithkode replied:
Thanks @sarayourfriend, your suggestion worked and the favicon now displays for individual API pages outside of DRF spectacular's scope. Sadly, still no luck getting the favicon to show up on the API index managed by spectacular using the `SWAGGER_UI_FAVICON_HREF` declaration. I have repeatedly cleared my browse cache, tried multiple browsers, played around with the url path to `favicon.ico` and none seemed to work. I have once again updated the MR with my current local state. Hoping this issue can be replicated by someone else.

madewithkode replied:
This makes sense. I'd try it out and revert, hoping it works!

krysal replied:
The "funny" thing is that the solution should be similar, modifying the [redoc template](https://github.com/tfranzel/drf-spectacular/blob/master/drf_spectacular/templates/drf_spectacular/redoc.html), that @dhruvkb asked to support templating last year! tfranzel/drf-spectacular#978 

It's not working for me tho, so the Nginx solution seems the most straightforward for now. We'll get back to this next week if you need more help.

sarayourfriend commented:
@dhruvkb this PR does not close https://github.com/WordPress/openverse/issues/4373. Did you intend to link them?

dhruvkb commented:
@sarayourfriend the change to `EXTENSIONS_INFO.x-logo.url` in `api/conf/settings/spectacular.py` should fix the logo in the sidebar.

sarayourfriend commented:
The issue's requirements are to host the image ourselves, rather than use GitHub, FWIW, but I suppose it would strictly solve the issue of the logo missing.

krysal commented on api/conf/settings/spectacular.py:
```diff
@@ -41,10 +42,11 @@
             "backgroundColor": "#fafafa",
         }
     },
+    "SWAGGER_UI_FAVICON_HREF": f"{STATIC_URL}favicon.ico",
```
Setting the URL from the root makes it work for swagger: http://localhost:50280/v1/swagger/#/ 
```suggestion
    "SWAGGER_UI_FAVICON_HREF": f"/{STATIC_URL}favicon.ico",
```
Although this does not solve the problem for the index.

dhruvkb replied:
Ah yes, I think its worth adding the leading slash to `STATIC_URL` itself in the conf, but that can be addressed separately.

https://github.com/WordPress/openverse/blob/18a6606cd5ef56d74d9b3f592e8c899148474aee/api/conf/settings/static.py#L10

madewithkode commented:
@dhruvkb @krysal It appears that doing this at the nginx proxy level as suggested by @sarayourfriend is so far the most consistent solution. Therefore, it's worth noting that accessing the api as served by django would likely not have favicons show up but accessing via the nginx proxy works fine. I have now updated this PR with that change.

dhruvkb reviewed (commented):
The SVG seems broken. I previewed it in https://www.svgviewer.dev/ and it only shows up as a white square with black border.

madewithkode commented:
@dhruvkb thanks for spotting the svg issue. Fixed that and updated the PR.

sarayourfriend reviewed (approved):
This is great, much simpler in the end!

dhruvkb reviewed (approved):
It doesn't work directly for the `web` service, which means I won't get to experience it locally 😢 but it will work on the actual website!
